fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateSplit.java

```diff
@@ -75,33 +88,196 @@ public static PredicateSplit splitPredicate(ScalarOperator predicate) {
+            Utils.compoundAnd(extractor.residualPredicates));
+    }
+
+    public static class RangeExtractorContext {
+        private boolean isAnd = true;
+
+        public boolean isAnd() {
+            return isAnd;
+        }
+
+        public RangeExtractorContext setAnd(boolean and) {
+            isAnd = and;
+            return this;
+        }
+    }
+
+    public static class RangeExtractor
+            extends ScalarOperatorVisitor<Map<ColumnRefOperator, TreeRangeSet<ConstantOperator>>, RangeExtractorContext> {
+
+        List<ScalarOperator> columnEqualityPredicates = Lists.newArrayList();
         List<ScalarOperator> residualPredicates = Lists.newArrayList();
-        // Split predicates into three kinds:
-        //  - Equal Predicates:  col1 = col2 (both col1 and col2 are ColumnRefs)
-        //  - Range Predicates:  col1 >/>=/=/</<= 2 (col1 is ColumnRef and right is ConstantRef)
-        // TODO: support NotEqual/Or Range as range predicates
-        //  - Other Predicates:  others(eg: NonEqual BinaryPredicateOperator or others)
-        for (ScalarOperator scalarOperator : predicateConjuncts) {
-            if (scalarOperator instanceof BinaryPredicateOperator) {
-                BinaryPredicateOperator binary = (BinaryPredicateOperator) scalarOperator;
-                ScalarOperator leftChild = scalarOperator.getChild(0);
-                ScalarOperator rightChild = scalarOperator.getChild(1);
-                BinaryType binaryType = binary.getBinaryType();
-                if (binaryType.isEqual() && leftChild.isColumnRef() && rightChild.isColumnRef()) {
-                    columnEqualityPredicates.add(scalarOperator);
-                } else if (binaryType.isEqualOrRange() && leftChild.isColumnRef() && rightChild.isConstantRef()) {
-                    rangePredicates.add(scalarOperator);
+
+        @Override
+        public Map<ColumnRefOperator, TreeRangeSet<ConstantOperator>> visit(
+                ScalarOperator scalarOperator, RangeExtractorContext context) {
+            if (scalarOperator instanceof PredicateOperator) {
+                return visitPredicate((PredicateOperator) scalarOperator, context);
+            }
+            return scalarOperator.accept(this, context);
+        }
+
+        @Override
+        public Map<ColumnRefOperator, TreeRangeSet<ConstantOperator>> visitPredicate(
+                PredicateOperator predicate, RangeExtractorContext context) {
+            return null;
+        }
+
+        @Override
+        public Map<ColumnRefOperator, TreeRangeSet<ConstantOperator>> visitBinaryPredicate(
+                BinaryPredicateOperator predicate, RangeExtractorContext context) {
+            ScalarOperator left = predicate.getChild(0);
+            ScalarOperator right = predicate.getChild(1);
+            if (left.isColumnRef() && right.isConstantRef()) {
+                if (predicate.getBinaryType().isNotEqual()) {
```

Comment:
still cannot change to `!=` to range predicates?

Reply:
already supported